"""
Demo Client - é€šç”¨å®¢æˆ¶ç«¯ç¤ºç¯„

å±•ç¤ºå¦‚ä½•ä½¿ç”¨ ServiceClient å’Œ ServiceRegistry 
èˆ‡ HTTP å’Œ gRPC æœå‹™é€²è¡Œäº¤äº’ã€‚
"""

import sys
from pathlib import Path

# Add parent directory to path
sys.path.insert(0, str(Path(__file__).parent))

from service_client import ServiceClient, ServiceRegistry
from generated.pose_detector_pb2 import DetectPoseRequest, DetectPoseResponse
from generated.pose_classifier_pb2 import ClassificationRequest, ClassificationResponse


def demo_simple_client():
    """ç¤ºç¯„ 1: ä½¿ç”¨ç°¡å–®çš„ ServiceClient"""
    print("=" * 70)
    print("  ç¤ºç¯„ 1: ç°¡å–®å®¢æˆ¶ç«¯ (ServiceClient)")
    print("=" * 70)
    
    # å‰µå»º HTTP å®¢æˆ¶ç«¯
    print("\nğŸ“¡ å‰µå»º HTTP å®¢æˆ¶ç«¯é€£æ¥åˆ° localhost:8081...")
    detector_client = ServiceClient("localhost:8081", protocol="http")
    print(f"âœ“ å®¢æˆ¶ç«¯å·²å‰µå»º: {detector_client}")
    
    # èª¿ç”¨æª¢æ¸¬æœå‹™
    print("\nğŸ“¸ èª¿ç”¨å§¿æ…‹æª¢æ¸¬æœå‹™...")
    request = DetectPoseRequest(frame_id="frame_demo_001")
    
    try:
        response = detector_client.call(
            path="/detect_pose",
            request=request,
            response_type=DetectPoseResponse
        )
        
        print("âœ“ æª¢æ¸¬æˆåŠŸ!")
        print(f"  - Frame ID: {request.frame_id}")
        print(f"  - Head: ({response.pose.head.x}, {response.pose.head.y}, {response.pose.head.z})")
        print(f"  - Torso: ({response.pose.torso.x}, {response.pose.torso.y}, {response.pose.torso.z})")
        
        return response
    except Exception as e:
        print(f"âœ— èª¿ç”¨å¤±æ•—: {e}")
        return None


def demo_registry():
    """ç¤ºç¯„ 2: ä½¿ç”¨ ServiceRegistry ç®¡ç†å¤šå€‹æœå‹™"""
    print("\n" + "=" * 70)
    print("  ç¤ºç¯„ 2: æœå‹™è¨»å†Šè¡¨ (ServiceRegistry)")
    print("=" * 70)
    
    # å‰µå»ºè¨»å†Šè¡¨
    print("\nğŸ“‹ å‰µå»ºæœå‹™è¨»å†Šè¡¨...")
    registry = ServiceRegistry()
    
    # è¨»å†Šæœå‹™
    print("ğŸ“ è¨»å†Šæœå‹™...")
    registry.register("detector", "localhost:8081", protocol="http")
    registry.register("classifier", "localhost:8082", protocol="http")
    
    print(f"âœ“ è¨»å†Šå®Œæˆ: {registry}")
    print(f"  å·²è¨»å†Šæœå‹™: {registry.list_services()}")
    
    # èª¿ç”¨æª¢æ¸¬æœå‹™
    print("\nğŸ“¸ æ­¥é©Ÿ 1: èª¿ç”¨å§¿æ…‹æª¢æ¸¬æœå‹™...")
    detect_request = DetectPoseRequest(frame_id="frame_demo_002")
    
    try:
        detect_response = registry.call(
            service_name="detector",
            path="/detect_pose",
            request=detect_request,
            response_type=DetectPoseResponse
        )
        
        print("âœ“ æª¢æ¸¬æˆåŠŸ!")
        print(f"  - Head: ({detect_response.pose.head.x}, {detect_response.pose.head.y}, {detect_response.pose.head.z})")
        print(f"  - Torso: ({detect_response.pose.torso.x}, {detect_response.pose.torso.y}, {detect_response.pose.torso.z})")
        
    except Exception as e:
        print(f"âœ— èª¿ç”¨å¤±æ•—: {e}")
        return
    
    # èª¿ç”¨åˆ†é¡æœå‹™
    print("\nğŸ” æ­¥é©Ÿ 2: èª¿ç”¨å§¿æ…‹åˆ†é¡æœå‹™...")
    classify_request = ClassificationRequest(pose=detect_response.pose)
    
    try:
        classify_response = registry.call(
            service_name="classifier",
            path="/classify_pose",
            request=classify_request,
            response_type=ClassificationResponse
        )
        
        print("âœ“ åˆ†é¡æˆåŠŸ!")
        print(f"  - æ˜¯å¦å€’ç«‹: {classify_response.is_upside_down}")
        print(f"  - åˆ¤æ–·: {'å€’ç«‹' if classify_response.is_upside_down else 'æ­£å¸¸'}")
        
    except Exception as e:
        print(f"âœ— èª¿ç”¨å¤±æ•—: {e}")


def demo_auto_detect():
    """ç¤ºç¯„ 3: è‡ªå‹•æª¢æ¸¬å”è­°"""
    print("\n" + "=" * 70)
    print("  ç¤ºç¯„ 3: è‡ªå‹•å”è­°æª¢æ¸¬")
    print("=" * 70)
    
    print("\nğŸ” å‰µå»ºå®¢æˆ¶ç«¯ä¸¦è‡ªå‹•æª¢æ¸¬å”è­°...")
    
    # è‡ªå‹•æª¢æ¸¬ HTTP æœå‹™
    client = ServiceClient("localhost:8081", protocol="auto")
    print(f"âœ“ è‡ªå‹•æª¢æ¸¬çµæœ: {client}")
    
    # æ¸¬è©¦èª¿ç”¨
    print("\nğŸ“ æ¸¬è©¦èª¿ç”¨æœå‹™...")
    request = DetectPoseRequest(frame_id="frame_auto_detect")
    
    try:
        response = client.call(
            path="/detect_pose",
            request=request,
            response_type=DetectPoseResponse
        )
        print(f"âœ“ èª¿ç”¨æˆåŠŸ! å”è­°: {client.protocol.value}")
    except Exception as e:
        print(f"âœ— èª¿ç”¨å¤±æ•—: {e}")


def demo_grpc_client():
    """ç¤ºç¯„ 4: gRPC å®¢æˆ¶ç«¯ (å¦‚æœæœ‰ gRPC æœå‹™é‹è¡Œ)"""
    print("\n" + "=" * 70)
    print("  ç¤ºç¯„ 4: gRPC å®¢æˆ¶ç«¯ (å¯é¸)")
    print("=" * 70)
    
    print("\nğŸ“¡ å‰µå»º gRPC å®¢æˆ¶ç«¯é€£æ¥åˆ° localhost:50051...")
    print("   (å¦‚æœæ²’æœ‰ gRPC æœå‹™é‹è¡Œï¼Œæ­¤ç¤ºç¯„æœƒè·³é)")
    
    try:
        client = ServiceClient("localhost:50051", protocol="grpc")
        print(f"âœ“ å®¢æˆ¶ç«¯å·²å‰µå»º: {client}")
        
        # èª¿ç”¨æœå‹™
        request = DetectPoseRequest(frame_id="frame_grpc_001")
        response = client.call(
            path="/detect_pose",
            request=request,
            response_type=DetectPoseResponse
        )
        
        print("âœ“ gRPC èª¿ç”¨æˆåŠŸ!")
        print(f"  - Head: ({response.pose.head.x}, {response.pose.head.y}, {response.pose.head.z})")
        
    except Exception as e:
        print(f"âš ï¸  gRPC æœå‹™æœªé‹è¡Œæˆ–èª¿ç”¨å¤±æ•—: {e}")
        print("   æç¤º: ä½¿ç”¨ protocol='grpc' å•Ÿå‹•æœå‹™ä¾†æ¸¬è©¦æ­¤åŠŸèƒ½")


def demo_comparison():
    """ç¤ºç¯„ 5: HTTP vs gRPC å°æ¯”"""
    print("\n" + "=" * 70)
    print("  ç¤ºç¯„ 5: å”è­°å°æ¯” (HTTP vs gRPC)")
    print("=" * 70)
    
    request = DetectPoseRequest(frame_id="frame_comparison")
    
    # HTTP
    print("\nğŸŒ HTTP å”è­°:")
    try:
        http_client = ServiceClient("localhost:8081", protocol="http")
        response = http_client.call("/detect_pose", request, DetectPoseResponse)
        print(f"  âœ“ æˆåŠŸ - å”è­°: HTTP")
    except Exception as e:
        print(f"  âœ— å¤±æ•—: {e}")
    
    # gRPC  
    print("\nâš¡ gRPC å”è­°:")
    try:
        grpc_client = ServiceClient("localhost:50051", protocol="grpc")
        response = grpc_client.call("/detect_pose", request, DetectPoseResponse)
        print(f"  âœ“ æˆåŠŸ - å”è­°: gRPC")
    except Exception as e:
        print(f"  âš ï¸  æœªé‹è¡Œæˆ–å¤±æ•—: {e}")
    
    print("\nğŸ“Š å”è­°ç‰¹é»å°æ¯”:")
    print("  HTTP:")
    print("    + ç°¡å–®æ˜“ç”¨ï¼Œå»£æ³›æ”¯æ´")
    print("    + é˜²ç«ç‰†å‹å¥½")
    print("    - æ•ˆèƒ½è¼ƒä½")
    print("  gRPC:")
    print("    + é«˜æ•ˆèƒ½ï¼Œä½å»¶é²")
    print("    + æ”¯æ´ä¸²æµ")
    print("    - éœ€è¦ç‰¹æ®Šé…ç½®")


def main():
    client = ServiceClient("localhost:8081", protocol="auto")
    
    request = DetectPoseRequest(frame_id="frame_auto_detect")
    
    try:
        response = client.call(
            path="/detect_pose",
            request=request,
            response_type=DetectPoseResponse
        )
        print(f"âœ“ èª¿ç”¨æˆåŠŸ! å”è­°: {client.protocol.value}")
    except Exception as e:
        print(f"âœ— èª¿ç”¨å¤±æ•—: {e}")

if __name__ == "__main__":
    main()
